babybuddy/templates/babybuddy/login_qr_code.txt

```diff
@@ -0,0 +1,4 @@
+{% load i18n widget_tweaks babybuddy_tags qr_code %}
+{% url 'babybuddy:root-router' as relative_root_url %}
+{% make_absolute_url relative_root_url as absolute_root_url %}
+BABYBUDDY-LOGIN:{"url":"{{ absolute_root_url }}","api_key":"{{ user.settings.api_key }}"}
```

Comment:
I wonder... is there any way we can make this _also_ work with a regular browser? Or is that not a thing? I'm unsure how QR codes work haha. It seems like it would be convenient as a way to easily log in via a mobile web browser from a logged in desktop, for example.

Based on this random page it seems like this should be possible but I can't find any documentation about how to have a URL "fallback": https://qr-creator.com/detect-iOS-Android.php

If that's not possible we'll need to think on a better way to explain what the code is for... maybe just calling the field "Android App login code"? Seems kind of verbose...

Reply:
Hey... took some time for me to write a complete answer on this topic specifically. It namely hits home for me wrt a problem I am facing with the Android app as well:

Right now the `api_key` actually only seems to give access to the `/api`-views. I think this is not too crazy, actually. The `api_key` is for using that API. What I am trying to do with the Android app is to make using timers very convenient and send everyone to the actual website for everything else. For this I need to forward people to the website, that I cannot show without a valid user session. User sessions cannot be generated with the `api_key`.

So the bottom-line seems to be: "Yes, this does not exist" for the regular website - unfortunately. At least not with the current authentication mechanisms.

--------------------

I really would like to do something like that in the future using the `api_key` for my app. So I was thinking myself about this. "Funnily" enough, I am maintaining a plugin for wordpress that does exactly that: https://github.com/MrApplejuice/wp-autologin-links (kind of dormat right now) - it adds a token-login (like the api-key is, in principle).

From my experience with the plugin, I would however suggest that not to do this. It is super-unsafe, generally. Typically one has to embed the login code in the url as a get-parameter, because there is no other safe way of embedding the code in a plain-text representation (like a QR-code is), but then again it is super easy to share such a link by accident for a user and pretty much leak his/her password and can get full access to his/her account. This is different from the API, which allows a lot of access, but not, for example, to change one's password. Such a code is convenient, but only if the website is designed for that. I do not know that this would be a "good feature" in that sense.

What I would suggest instead is to create a special backend on the API-side that will allow generating actual user sessions using a POST message, for example. This will open up an avenue for developers to open user sessions without storing a user password for authentication purposes. Since we are talking about POST-requests, the secret `api_url` can remain a header and cannot be encoded in the URL anyway (security feature), and it would be possible to configure that special endpoint in such a way that allows admins to lock down that potential vulnerability if is undesirable.

I think this is my stance on this. Token authentication is handy but not super-safe. There are better ways that integrators can use and users should probably stick to password authentication and store their passwords in a password manager which is just as convenient as a login-token.

Reply:
> It is super-unsafe, generally. 

Yeah. I tend to think that's OK for an app like this but I suppose that's definitely not true of all users (or even a majority, who knows!). So maybe this would just be a nice optional feature for some (using e.g. https://github.com/aaugustin/django-sesame) -- separately.

> What I would suggest instead is to create a special backend on the API-side that will allow generating actual user sessions using a POST message, for example. This will open up an avenue for developers to open user sessions without storing a user password for authentication purposes.

Whats the use case for this? I'm unclear on that.

Reply:
Regardless I still want to make clear to the user what this QR code is for. As it currently it the "Login QR code" title is kind of confusing. Maybe we should generate the code on a separate page and link to it from somewhere instead? Or put this in a separate "Android" fieldset?

Reply:
@MrApplejuice any thoughts on follow up here? I can fiddling with this if you've got more pressing issues right now 👶

Reply:
Hey... whoops, sorry. Somehow I marked this as "done" in my head. 

> Regardless I still want to make clear to the user what this QR code is for. As it currently it the "Login QR code" title is kind of confusing.

"Hiding" the page somewhere else could indeed be a nice option. I quickly checked how Telegram does it which I had ready on my phone. It has a menu-item "Link a new device" under the menu option "Devices". Not really something that babybuddy can copy 1:1 because it would mean that individual devices get their own API keys (though that would be cool, I guess).

Maybe something like this (quick mock-up):

![image](https://user-images.githubusercontent.com/542105/194139534-ab371c26-b540-4910-ab3d-b85135ac5eb8.png)

And then on the new page you get presented with the API-URL and the QR-code maybe?

Reply:
> Hey... whoops, sorry. Somehow I marked this as "done" in my head.

Haha I did too.

> Maybe something like this (quick mock-up):

That looks fine -- lets do it!

Reply:
Quick expectation management: I will need another week to implement this. Hope that is no problem?!

Reply:
Absolutely. Baby Buddy has no deadlines 😀